File: notes_and_examples/overloading_overriding.py
"""

In Python, method overriding is a concept used in inheritance. 
When a method in a child class has the same name and signature as a method in the parent class, 
it overrides the method in the parent class.


"""

class Animal:
    def make_sound(self):
        print("Generic sound")

class Dog(Animal):
    def make_sound(self):
        print("Bark")

# Usage
animal = Animal()
animal.make_sound()  # Output: Some generic sound

dog = Dog()
dog.make_sound()  # Output: Bark


"""

In Python, overloading isn't directly supported as in languages like Java or C++. 
However, Python allows function overloading by default due to its dynamic nature. 
Functions can be defined in a way that they behave differently based on the number or types of arguments passed.

"""

class OverloadExample:
    def add(self, a, b):
        return a + b

    def add(self, a, b, c):
        return a + b + c

# Usage
overload = OverloadExample()
print(overload.add(2, 3))        # This would raise an error as Python does not support traditional method overloading.
print(overload.add(2, 3, 4))    # This will work as expected.

File: notes_and_examples/overloading_overriding.java
// In Java, method overloading allows a class to have more than one method with the same name if their parameter lists are different.

public class OverloadExample {
    public int add(int a, int b) {
        return a + b;
    }

    public int add(int a, int b, int c) {
        return a + b + c;
    }

    public static void main(String[] args) {
        OverloadExample overload = new OverloadExample();
        System.out.println(overload.add(2, 3));        // Output: 5
        System.out.println(overload.add(2, 3, 4));    // Output: 9
    }
}


// In this example, the add method is overloaded with two versions having different numbers of parameters. 
// Depending on the arguments passed, the appropriate method is called.

// In Java, method overriding occurs when a subclass provides a specific implementation of a method that is already provided by its superclass.

class Animal {
    public void makeSound() {
        System.out.println("Generic sound");
    }
}

class Dog extends Animal {
    @Override
    public void makeSound() {
        System.out.println("Bark");
    }
}

public class OverrideExample {
    public static void main(String[] args) {
        Animal animal = new Animal();
        animal.makeSound();  // Output: Generic sound

        Animal dog = new Dog();
        dog.makeSound();     // Output: Bark 
    }
}



